Use EVP_Digest APIs for s2n_hash when in FIPS mode

bpdavidson opened the pull request:
Note: These changes currently break our SAW tests. Galois is working on changes to the SAW tests that will be included in this PR before it is merged.

As discussed in Issue #478, s2n_hash must use OpenSSL's EVP_Digest APIs when in FIPS mode with a FIPS-capable libcrypto. These APIs will only be used when in FIPS mode currently as there is request path memory allocation within the EVP_DigestInit and EVP_DigestCopy APIs.

bpdavidson commented on crypto/s2n_hash.h:
```diff
@@ -63,13 +66,34 @@ struct s2n_hash_state {
+    } low_level;
+    struct s2n_hash_evp_digest {
+        struct s2n_evp_digest primary;
+        struct s2n_evp_digest md5_sha1_secondary;
```
Agreed and updated.

bpdavidson commented on crypto/s2n_hmac.c:
```diff
@@ -258,6 +275,18 @@ int s2n_hmac_digest_verify(const void *a, const void *b, uint32_t len)
+    GUARD(s2n_hash_copy(&to->inner_just_key, &from->inner_just_key));
+    GUARD(s2n_hash_copy(&to->outer, &from->outer));
+
+    memcpy_check(to->xor_pad, from->xor_pad, 128*sizeof(uint8_t));
```
Fixed.

bpdavidson commented on crypto/s2n_hmac.c:
```diff
@@ -258,6 +275,18 @@ int s2n_hmac_digest_verify(const void *a, const void *b, uint32_t len)
+    GUARD(s2n_hash_copy(&to->outer, &from->outer));
+
+    memcpy_check(to->xor_pad, from->xor_pad, 128*sizeof(uint8_t));
+    memcpy_check(to->digest_pad, from->digest_pad, SHA512_DIGEST_LENGTH*sizeof(uint8_t));
```
Fixed.

alexeblee commented on crypto/s2n_ecdsa.c:
```diff
@@ -157,26 +161,26 @@ int s2n_ecdsa_signature_size(const struct s2n_ecdsa_private_key *key)
 
-int s2n_ecdsa_keys_match(const struct s2n_ecdsa_public_key *pub_key, const struct s2n_ecdsa_private_key *priv_key) 
+int s2n_ecdsa_keys_match(const struct s2n_ecdsa_public_key *pub_key, struct s2n_hash_state *state_in,
+                         const struct s2n_ecdsa_private_key *priv_key, struct s2n_hash_state *state_out)
```
I don't like how this signature is now exposing the internals of how the function works. It feels to me like a slight violation of encapsulation. To me, the caller should just be able to call a function that says, "given this key and this other key, do they match?" It shouldn't need to provide a hash algorithm to be able to help prove it. Is this just to try to avoid having to reallocate hash state?

bpdavidson replied:
I agree. Yes, this change was intended to avoid any request path allocation of s2n_hash_states (as you can see in the ECDSA unit test where they are allocated before this call). Given your work in this ECDSA area, do you have any thoughts on how to avoid this allocation?

alexeblee commented on crypto/s2n_hash.c:
```diff
@@ -169,13 +173,270 @@ int s2n_hash_digest(struct s2n_hash_state *state, void *out, uint32_t size)
+static int s2n_evp_hash_new(struct s2n_hash_state *state)
+{
+    notnull_check(state->digest.evp.primary.ctx = S2N_EVP_MD_CTX_NEW());
+    notnull_check(state->digest.evp.md5_secondary.ctx = S2N_EVP_MD_CTX_NEW());
```
Hmm.. Is this md5_secondary only used in the case of md5_sha1? Is there any way to make this cleaner so we don't have to have a default secondary since it won't be used for all cases?

bpdavidson replied:
Yes, it's only used for md5_sha1. This can't be much cleaner as we want to preserve s2n_hash_state's ability to be reset/init'ed to a different hash algorithm without additional allocation required. This allows us to pre-allocate all memory needed for all s2n_hash_algorthms.

alexeblee commented on crypto/s2n_hash.c:
```diff
@@ -169,13 +173,270 @@ int s2n_hash_digest(struct s2n_hash_state *state, void *out, uint32_t size)
+    case S2N_HASH_SHA224:
+    case S2N_HASH_SHA256:
+    case S2N_HASH_SHA384:
+    case S2N_HASH_SHA512:
```
I thought I remembered you saying that you were going to use a wrapper-style interface (similar to how s2n_cipher is) instead of the switch-case-style. Was that still something you were considering? I personally feel like the wrapper-style is cleaner and more consistent with other places in the code base, but I could be wrong.

bpdavidson replied:
The wrapper interface that I was referencing was around low_level vs high_level hash, to keep callers blind to the implementation used by preserving the s2n_hash_<func> calls as-is. A per-algorithm wrapper may provide some benefit, but would need to be constructed such that an s2n_hash_state can be reset to a different algorithm without additional allocation. Ultimately, I think that a per-algorithm wrapper implementation is outside of the scope of the requirements that this PR is targeting (FIPS compliance).

alexeblee commented on crypto/s2n_hash.c:
```diff
@@ -169,13 +173,270 @@ int s2n_hash_digest(struct s2n_hash_state *state, void *out, uint32_t size)
+    .digest = &s2n_low_level_hash_digest,
+    .copy = &s2n_low_level_hash_copy,
+    .reset = &s2n_low_level_hash_reset,
+    .free = &s2n_low_level_hash_free,
```
Related to my previous comment - I think you could create a different struct of function pointers for each hash type that called the algorithm-specific version of the function directly instead of the s2n_low_level_hash* generic function that then calls the algorithm-specific version.

bpdavidson replied:
(See reply above)

alexeblee commented on crypto/s2n_hmac.c:
```diff
@@ -258,6 +275,18 @@ int s2n_hmac_digest_verify(const void *a, const void *b, uint32_t len)
+    to->currently_in_hash_block = from->currently_in_hash_block;
+    to->block_size = from->block_size;
+    to->digest_size = from->digest_size;
+
```
This may be a dumb question, but why did we have to switch from memcpy?

bpdavidson replied:
Good question! OpenSSL's EVP_Digest API's require (and enforce) the use of their copy API to copy any hash state. They may have some external state that the API is also preserving. I have added a comment to ensure that we remember this.

alexeblee commented on tls/s2n_connection_handles.c:
```diff
@@ -0,0 +1,236 @@
+    /* Preserve handlers for TLS PRF p_hash to avoid re-allocation */
+    prf_handles->p_hash_s2n_hmac_inner = conn->prf_space.tls.p_hash.s2n_hmac.inner.digest.evp;
+    prf_handles->p_hash_s2n_hmac_inner_impl = conn->prf_space.tls.p_hash.s2n_hmac.inner.hash_impl;
+    prf_handles->p_hash_s2n_hmac_inner_just_key = conn->prf_space.tls.p_hash.s2n_hmac.inner_just_key.digest.evp;
```
Could some of this be stored in an intermediate variable to cut down on the accessor chain? e.g. conn->prf_space.tls.p_hash.s2n_hmac

bpdavidson replied:
I have pulled in changes from what was going to be another PR (#547) to add global state on which hash implementation to use rather than this per-hash state. That change eliminates the need to preserve any hash_impl's on s2n_connection_wipe.

alexeblee commented on tls/s2n_connection_handles.c:
```diff
@@ -0,0 +1,236 @@
+{
+    /* Preserve handlers for handshake hash states to avoid re-allocation */
+    hash_handles->md5 = conn->handshake.md5.digest.evp;
+    hash_handles->md5_hash_impl = conn->handshake.md5.hash_impl;
```
Do all of the hash_impl's need to be "preserved" or could they later be inferred and assigned? I thought they were just the function pointers and weren't something that was allocated per connection?

bpdavidson replied:
I have pulled in changes from what was going to be another PR (#547) to add global state on which hash implementation to use rather than this per-hash state. That change eliminates the need to preserve any hash_impl's on s2n_connection_wipe.

alexeblee commented on tls/s2n_prf.c:
```diff
@@ -196,39 +200,41 @@ static int s2n_evp_hmac_p_hash_free(union s2n_prf_working_space *ws)
-    return 0;
+    GUARD(s2n_hmac_new(&ws->tls.p_hash.s2n_hmac));
+
+    return s2n_hmac_init(&ws->tls.p_hash.s2n_hmac, S2N_HMAC_NONE, NULL, 0);
```
Could you give me the quick explanation as to why this was added?

bpdavidson replied:
s2n_hmac_p_hash_new is called via s2n_prf_new() when out of FIPS mode and thus, is using s2n's HMAC implementation for the p_hash. 

Currently, if s2n is not in FIPS mode, s2n_hmac_new uses the low_level s2n_hash and this is a no-op. If we change our logic such that s2n_hash always uses the high_level EVP's, this will not be a no-op. So, it's just ensuring that we are using the s2n_hmac API's in a consistent manor to prevent future changes from breaking us.

bpdavidson commented on crypto/s2n_hash.h:
```diff
@@ -63,13 +66,34 @@ struct s2n_hash_state {
         } md5_sha1;
-    } hash_ctx;
+    } low_level;
+    struct s2n_hash_evp_digest {
```
I have re-worked the naming a bit in this most recent revision to address your comments. Let me know if this helps.

bpdavidson commented on crypto/s2n_hash.h:
```diff
@@ -63,13 +66,34 @@ struct s2n_hash_state {
+    union s2n_hash_digest digest;
+};
+
+struct s2n_hash {
```
I have re-worked the structure of s2n_hash_state, removed the internal use of s2n_hash (function pointers) and abstracted that to a global variable, and re-worked naming. Let me know if that helps.

bpdavidson commented on tls/s2n_crypto.h:
```diff
@@ -77,7 +77,9 @@ struct s2n_crypto_parameters {
 
+    struct s2n_hash_state signature_hash;
     struct s2n_hmac_state client_record_mac;
+    struct s2n_hmac_state client_record_mac_copy;
```
s2n_verify_cbc uses an hmac copy that needs to be allocated and freed at the s2n_connection level. I renamed this to record_mac_copy_workspace for clarity.

bpdavidson commented on crypto/s2n_hash.c:
```diff
@@ -39,34 +38,39 @@ int s2n_hash_digest_size(s2n_hash_algorithm alg, uint8_t *out)
-int s2n_hash_init(struct s2n_hash_state *state, s2n_hash_algorithm alg)
+static int s2n_low_level_hash_new(struct s2n_hash_state *state)
+{
+    return 0;
```
Good call - added.

bpdavidson commented:
@alexeblee In regards to your comment:
> Short of stuffing the hash_state into the key, I don't know if there is an elegant way to avoid allocation. I don't think, however, that this function is called per-request or per-connection. So far, it's use is limited to s2n_config setup when configuring certs. If that is something we can live with, that'd make this a lot simpler. If not, I'm okay with keeping it as-is and addressing it when we refactor the ecdsa functions (which is currently in progress) as we discussed offline.

I implemented your proposal of allowing allocation for s2n_hash_state within s2n_ecdsa_keys_match. As you mentioned, there is no plan to use this on a per-request or per-connection basis. In addition, s2n_hash_new does not allocate any memory when outside of FIPS mode. I have added a comment in s2n_ecdsa_keys_match noting that memory allocation in s2n_hash_new is currently restricted to FIPS mode. If you find that s2n_ecdsa_keys_match will be implemented on a per-request or per-connection basis, we should revisit this decision.

bpdavidson commented on crypto/s2n_hash.c:
```diff
@@ -205,26 +211,26 @@ static int s2n_evp_hash_init(struct s2n_hash_state *state, s2n_hash_algorithm al
         break;
     case S2N_HASH_MD5:
-        r = EVP_DigestInit_ex(state->digest.evp.primary.ctx, EVP_md5(), NULL);
+        r = EVP_DigestInit_ex(state->digest.high_level.evp.ctx, EVP_md5(), NULL);
```
I agree, good suggestion. I've done this for all high-level implementation s2n_hash calls where there is a common EVP_MD_CTX in each case of the switch statement.

bpdavidson replied:
I had made these changes but after rebasing my s2n FIPS unit test changes onto this branch, these changes broke the build.

Due to OpenSSL's internal use of EVP_MD_CTX when in FIPS mode, the EVP_Digest_copy API must be used to copy pointers to an EVP_MD_CTX. For this reason, we cannot abstract the accessors within high-level hash functions.

I needed to revert these changes in the latest commit.

alexw91 commented on crypto/s2n_hash.c:
```diff
@@ -169,13 +176,301 @@ int s2n_hash_digest(struct s2n_hash_state *state, void *out, uint32_t size)
+    /* Ensure that hash_impl is set, as it may have been reset for s2n_hash_state on s2n_connection_wipe.
+     * When in FIPS mode, the EVP API's must be used for hashes.
+     */
+    s2n_is_in_fips_mode() ? (state->hash_impl = &s2n_evp_hash) : (state->hash_impl = &s2n_low_level_hash);
```
This line is the same as line 420 above and line 473 below, should we have a `s2n_set_hash_impl(struct s2n_hash_state *state)` function that's called in both places?

bpdavidson replied:
Agreed - added `s2n_hash_set_impl` (moved 'set' to conform with other s2n_hash function names).

alexw91 commented on tests/unit/s2n_hash_test.c:
```diff
@@ -26,85 +26,171 @@
+        EXPECT_SUCCESS(s2n_stuffer_write_uint8_hex(&output, digest_pad[i]));
+    }
+
+    /* Reference value from command line sha384sum */
```
sha512sum?

bpdavidson replied:
Yup, copy paste error. Fixed.

alexw91 commented on tests/unit/s2n_record_size_test.c:
```diff
@@ -87,7 +87,6 @@ int main(int argc, char **argv)
     EXPECT_SUCCESS(conn->secure.cipher_suite->record_alg->cipher->destroy_key(&conn->secure.server_key));
     EXPECT_SUCCESS(conn->secure.cipher_suite->record_alg->cipher->destroy_key(&conn->secure.client_key));
     EXPECT_SUCCESS(s2n_connection_free(conn));
-    EXPECT_SUCCESS(s2n_hmac_init(&conn->secure.server_record_mac, S2N_HMAC_SHA1, mac_key, sizeof(mac_key)));
```
Was this removed because the API Changed? Should we add back an equivalent check?

bpdavidson replied:
This must have been removed by mistake. The API changes do not affect this, so I've re-added it.

alexw91 commented on tls/s2n_handshake.h:
```diff
@@ -53,6 +53,13 @@ struct s2n_handshake {
+    struct s2n_hash_state md5_sha1;
+
+    /* Used for SSLv3 */
+    struct s2n_hash_state sslv3_md5_copy;
```
Where are these `sslv3_***_copy` Hash States used? Is that in a later commit? Same question for the `tls_hash_copy`.

bpdavidson replied:
These are used to copy the handshake state's hash being used to a copy workspace used by the PRF. I've added "PRF" to those comments for clarity.

alexw91 commented on crypto/s2n_hash.c:
```diff
@@ -169,13 +176,301 @@ int s2n_hash_digest(struct s2n_hash_state *state, void *out, uint32_t size)
+        r = EVP_DigestFinal_ex(primary_ctx, out, &digest_size);
+        break;
+    case S2N_HASH_MD5_SHA1:
+        r = EVP_DigestFinal_ex(primary_ctx, ((uint8_t *) out) + MD5_DIGEST_LENGTH, &digest_size);
```
We should create a new variable `int primary_digest_size = digest_size - MD5_DIGEST_LENGTH;` that can be used in the first `EVP_DigestFinal_ex` call for MD5_SHA1.

bpdavidson replied:
Great find! So, this is only going to affect the `s2n_evp_hash_digest` function (which calls OpenSSL's `EVP_DigestFinal_ex` API). In our other uses, size is either the size of the input data (for `s2n_hash_update`) or is not used within the OpenSSL API's (such as `s2n_low_level_hash_digest`).

I have modified this such that both calls to `EVP_DigestFinal_ex` for `S2N_HASH_MD5_SHA1` use the exact MD5/SHA1 digest size.

bpdavidson commented on crypto/s2n_fips.c:
```diff
@@ -17,16 +17,35 @@
+
+int s2n_fips_cleanup(void)
+{
+    s2n_fips_mode = 0;
```
Originally, the intention was to ensure that s2n_init was always initialized with s2n_fips_mode equal to 0. I have removed s2n_fips_cleanup and have modified s2n_fips_init to always set s2n_fips_mode to 0 if FIPS_mode is not set.

bpdavidson commented on crypto/s2n_fips.c:
```diff
@@ -17,16 +17,35 @@
+/* Return 1 if FIPS mode is enabled, 0 otherwise. FIPS mode must be enabled prior to calling s2n_init(). */
+int s2n_is_in_fips_mode(void)
+{
+    if (s2n_fips_mode) {
```
Good catch, changed.

alexw91 commented:
Looks Good To Me
